proxy/agent.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"reflect"
+	"time"
+
+	"k8s.io/client-go/pkg/util/flowcontrol"
+
+	"github.com/golang/glog"
+)
+
+// Agent manages the restarts and the life cycle of a proxy binary.  Agent
+// keeps track of all running proxy processes and their configurations.  Hot
+// restarts are performed by launching a new Envoy process with a strictly
+// incremented restart epoch.  This matches Envoy semantics for restart epochs:
+// To successfully launch a new Envoy process that will replace the running
+// Envoy processes, the restart epoch of the new process must be exactly 1
+// greater than the highest restart epoch of the currently running Envoy
+// processes. The initial epoch is 0.
+//
```

Comment:
Alternatively, we could rename the epoch to version (config file version [not schema version]). This would then be generic to different proxies